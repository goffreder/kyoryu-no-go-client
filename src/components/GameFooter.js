import React, { Component } from 'react';

import Captured from './Captured';
import ActiveColor from './ActiveColor';
import BoardMessage from './BoardMessage';
import PassButton from './PassButton';
import ResetBoardButton from './ResetBoardButton';
import QuitGameButton from './QuitGameButton';

export default class extends Component {
    render() {
        return (
            <div
                id="game-footer"
                className="page-footer blue-grey darken-2 row"
            >
                <div className="section col s6 m3">
                    <PassButton />
                </div>
                <div className="section section-vertical col s6 m1">
                    <ActiveColor />
                    <BoardMessage />
                </div>
                <div className="section col s12 m4">
                    <Captured />
                </div>
                <div className="section col s12 m4">
                    <ResetBoardButton />
                    <QuitGameButton />
                </div>
            </div>
        );
    }
}
