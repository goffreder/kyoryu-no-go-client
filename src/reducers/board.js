import isEmpty from 'lodash/isEmpty';
import includes from 'lodash/includes';

import { board as constants, msg } from '../constants';

export const defaultState = {
    board: [],
    captured: {
        [constants.BLACK]: 0,
        [constants.WHITE]: 0,
    },
    color: constants.BLACK,
    atari: false,
    suicide: false,
    passed: false,
    gameOver: false,
};

const reducer = {
    INIT_BOARD: (state, { payload }) => {
        return {
            ...state,
            board: Array(payload).fill(Array(payload).fill(constants.EMPTY)),
        };
    },
    RESET_BOARD: state => {
        return {
            ...defaultState,
            board: Array(state.board.length).fill(
                Array(state.board.length).fill(constants.EMPTY),
            ),
        };
    },
    QUIT_GAME: () => {
        return {
            ...defaultState,
        };
    },
    PLAY: (state, { payload: { row, col, color } }) => {
        if (state.gameOver) {
            return state;
        }

        if (
            state.board[row] === undefined ||
            state.board[row][col] !== constants.EMPTY
        ) {
            return state;
        }

        const activeColor = color || state.color;

        let newBoard = setBoardCell(state.board, row, col, activeColor);

        let neighbors = getAdjacentIntersections(state.board, row, col);
        let captured = [];
        let capturedStones = [];
        let atari = false;

        neighbors.forEach(n => {
            const i = state.board[n[0]][n[1]];

            if (
                i !== constants.EMPTY &&
                i !== activeColor &&
                !includes(capturedStones, `${n[0]},${n[1]}`)
            ) {
                const group = getGroup(newBoard, n[0], n[1]);

                if (group.liberties === 0) {
                    capturedStones = [
                        ...capturedStones,
                        ...group.stones.map(([r, c]) => `${r},${c}`),
                    ];
                    captured.push(group);
                }

                if (group.liberties === 1) {
                    atari = true;
                }
            }
        });

        if (isEmpty(captured) && getGroup(newBoard, row, col).liberties === 0) {
            return {
                ...state,
                suicide: true,
            };
        }

        captured.forEach(group => {
            group.stones.forEach(([row, col]) => {
                newBoard = setBoardCell(newBoard, row, col, constants.EMPTY);
            });
        });

        const newColor =
            activeColor === constants.BLACK ? constants.WHITE : constants.BLACK;

        return {
            ...state,
            board: newBoard,
            captured: {
                ...state.captured,
                [newColor]: state.captured[newColor] + capturedStones.length,
            },
            color: newColor,
            atari,
            suicide: false,
            passed: false,
        };
    },
    PASS: (state, { payload }) => {
        let newColor = null;

        if (state.passed) {
            newColor = state.color;
        } else {
            if (state.color === constants.BLACK) {
                newColor = constants.WHITE;
            } else {
                newColor = constants.BLACK;
            }
        }

        return {
            ...state,
            passed: true,
            color: newColor,
            gameOver: state.passed && true,
        };
    },
};

export const isBoardInit = state => !isEmpty(state.board);
export const getBoard = state => state.board;
export const getCaptured = state => state.captured;
export const getActiveColor = state => state.color;

export const getBoardMessage = state => {
    if (state.gameOver) {
        return msg.GAME_OVER;
    }

    if (state.suicide) {
        return msg.SUICIDE;
    }

    if (state.atari) {
        return msg.ATARI;
    }

    return '';
};

const setBoardCell = (board, row, col, color) => {
    const newRow = [
        ...board[row].slice(0, col),
        color,
        ...board[row].slice(col + 1, board[row].length),
    ];

    return [
        ...board.slice(0, row),
        newRow,
        ...board.slice(row + 1, board.length),
    ];
};

export const getAdjacentIntersections = (board, i, j) => {
    let neighbors = [];

    if (i > 0) {
        neighbors.push([i - 1, j]);
    }
    if (i < board.length - 1) {
        neighbors.push([i + 1, j]);
    }
    if (j < board[0].length - 1) {
        neighbors.push([i, j + 1]);
    }
    if (j > 0) {
        neighbors.push([i, j - 1]);
    }

    return neighbors;
};

export const getGroup = (board, i, j) => {
    const cellState = board[i][j];

    if (cellState === constants.EMPTY) {
        return {};
    }

    let visited = {};
    let visitedList = [];
    let queue = [[i, j]];
    let liberties = 0;

    while (queue.length) {
        const stone = queue.pop();

        if (visited[stone]) {
            continue;
        }

        const neighbors = getAdjacentIntersections(board, stone[0], stone[1]);

        for (let k = 0; k < neighbors.length; k++) {
            const n = neighbors[k];

            if (board[n[0]][n[1]] === constants.EMPTY) {
                liberties++;
            }
            if (board[n[0]][n[1]] === cellState) {
                queue.push([n[0], n[1]]);
            }
        }

        visited[stone] = true;
        visitedList.push(stone);
    }

    return {
        stones: visitedList,
        liberties,
    };
};

export default (state = defaultState, action = {}) => {
    if (typeof reducer[action.type] === 'undefined') {
        return state;
    }

    return reducer[action.type](state, action);
};
