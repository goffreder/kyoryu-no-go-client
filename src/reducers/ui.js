export const defaultState = {
    message: null,
};

const reducer = {
    SET_MESSAGE: (state, { payload }) => {
        return {
            ...state,
            message: payload,
        };
    },
};

export const getMessage = state => state.message;

export default (state = defaultState, action = {}) => {
    if (typeof reducer[action.type] === 'undefined') {
        return state;
    }

    return reducer[action.type](state, action);
};
