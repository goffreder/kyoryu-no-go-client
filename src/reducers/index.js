import { combineReducers } from 'redux';

import board, { defaultState as boardState } from './board';

export const defaultState = {
    board: {
        ...boardState,
    },
};

export default combineReducers({
    board,
});
