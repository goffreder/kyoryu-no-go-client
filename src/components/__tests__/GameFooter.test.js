import React from 'react';
import renderer from 'react-test-renderer';
import { createStore } from 'redux';
import { Provider } from 'react-redux';
import reducers, { defaultState } from '../../reducers';

import GameFooter from '../GameFooter';

describe('GameFooter component', () => {
    it('displays the game footer', () => {
        expect(
            renderer
                .create(
                    <Provider store={createStore(reducers, defaultState)}>
                        <GameFooter />
                    </Provider>,
                )
                .toJSON(),
        ).toMatchSnapshot();
    });
});
