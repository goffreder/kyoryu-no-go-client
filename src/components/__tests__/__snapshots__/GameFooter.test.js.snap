// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GameFooter component displays the game footer 1`] = `
<div
  className="page-footer blue-grey darken-2 row"
  id="game-footer"
>
  <div
    className="section col s6 m3"
  >
    <button
      className="waves-effect waves-light btn blue-grey darken-4"
      id="pass-btn"
      onClick={[Function]}
    >
      Pass
    </button>
  </div>
  <div
    className="section section-vertical col s6 m1"
  >
    <div
      id="active-color"
      style={
        Object {
          "height": 36,
          "position": "relative",
          "width": 36,
        }
      }
    >
      <div
        className="stone stone-black"
        style={
          Object {
            "background": "#101015",
            "borderRadius": 9001,
            "bottom": "5%",
            "boxShadow": Array [
              "inset 0 3.3333333333333335px 6.666666666666667px 0 rgba(255, 255, 255, .2)",
              "0 1.6666666666666667px 3.3333333333333335px 0 rgba(0, 0, 0, .16)",
              "0 1.6666666666666667px 5px 0 rgba(0, 0, 0, .12)",
            ],
            "left": "5%",
            "position": "absolute",
            "right": "5%",
            "top": "5%",
          }
        }
      />
    </div>
    <div
      id="alerts"
    >
      
    </div>
  </div>
  <div
    className="section col s12 m4"
  >
    <div
      id="captured-stones"
    >
      <div
        className="captured-wrapper"
      >
        <div
          style={
            Object {
              "height": 36,
              "position": "relative",
              "width": 36,
            }
          }
        >
          <div
            className="stone stone-white"
            style={
              Object {
                "background": "#EEEEF0",
                "borderRadius": 9001,
                "bottom": "5%",
                "boxShadow": Array [
                  "inset 0 -3.3333333333333335px 6.666666666666667px 0 rgba(0, 0, 0, .16)",
                  "0 1.6666666666666667px 3.3333333333333335px 0 rgba(0, 0, 0, .16)",
                  "0 1.6666666666666667px 5px 0 rgba(0, 0, 0, .12)",
                ],
                "left": "5%",
                "position": "absolute",
                "right": "5%",
                "top": "5%",
              }
            }
          />
        </div>
        <span>
          0
        </span>
      </div>
      <div
        className="captured-wrapper"
      >
        <div
          style={
            Object {
              "height": 36,
              "position": "relative",
              "width": 36,
            }
          }
        >
          <div
            className="stone stone-black"
            style={
              Object {
                "background": "#101015",
                "borderRadius": 9001,
                "bottom": "5%",
                "boxShadow": Array [
                  "inset 0 3.3333333333333335px 6.666666666666667px 0 rgba(255, 255, 255, .2)",
                  "0 1.6666666666666667px 3.3333333333333335px 0 rgba(0, 0, 0, .16)",
                  "0 1.6666666666666667px 5px 0 rgba(0, 0, 0, .12)",
                ],
                "left": "5%",
                "position": "absolute",
                "right": "5%",
                "top": "5%",
              }
            }
          />
        </div>
        <span>
          0
        </span>
      </div>
    </div>
  </div>
  <div
    className="section col s12 m4"
  >
    <button
      className="waves-effect waves-light btn blue-grey darken-4"
      id="reset-board-btn"
      onClick={[Function]}
    >
      Reset board
    </button>
    <button
      className="waves-effect waves-light btn blue-grey darken-4"
      id="quit-game-btn"
      onClick={[Function]}
    >
      Quit game
    </button>
  </div>
</div>
`;
